File: python/transformer.py
import yaml
import json
import getopt
import sys
import io

# Originally, this program would take a JSON file of an array of elements, capitalise them and write them to another file
# then we decided to add YAML support
# then somebody suggested it would be great to be able to take STDIN and emit to STDOUT, so we did that too
# Then somebody asked for lower-case as well. No problem!


def transform(filename, stream, input_format, output_format, transformation, output_file, output_stream):
    stuff = []
    raw_stuff = ''
    if filename is not None:
        fp = open(filename)
        if input_format is None:
            if filename.endswith(".json"):
                input_format = "JSON"
            elif filename.endswith(".yaml") or filename.endswith(".yml"):
                input_format = "YAML"
            if input_format is None:
                fp.close()
                raise Exception("No input format was specified")
        if input_format == 'YAML':
            stuff = yaml.safe_load(fp)
        else:
            raw_stuff = fp.read()
            stuff = json.loads(raw_stuff)
        fp.close()
    else:
        raw_stuff = stream.read()
        if input_format == None:
            raise Exception("No input format was specified")
        if input_format == "YAML":
            stuff = yaml.load(stuff)
        elif input_format == "JSON":
            stuff = json.load(raw_stuff)

    if output_format == None:
        if output_file.endswith(".json"):
            output_format = "JSON"
        elif output_file.endswith(".yaml") or output_file.endswith(".yml"):
            output_format = "YAML"
        else:
            output_format = input_format
    more_stuff = []

    if transformation == "CAPITALISE":
        for element in stuff:
            more_stuff.append(element.upper())
    elif transformation == "DECAPITALISE":
        for element in stuff:
            more_stuff.append(element.lower())

    if output_format == "JSON":
        if output_file is not None:
            json.dump(more_stuff, open(output_file, "w"))
            # TODO: shit, how to close this open()?
        elif output_stream is not None:
            json.dump(more_stuff, output_stream)
        else:
            raise Exception("No output stream is specified")

    if output_format == "YAML":
        if output_stream is not None:
            yaml.dump(more_stuff, output_stream)
        else:
            with open(output_file, "w") as fp:
                yaml.dump(more_stuff, fp)

    return


def main():
    options, arguments = getopt.getopt(
        sys.argv[1:],                      # Arguments
        'i:o:f:F:ul',                            # Short option definitions
        ["input", "output", "format", "output_format", "upper", "lower"])  # Long option definitions
    filename = None
    input_stream = None
    input_format = "JSON"
    output_file = None
    output_stream = None
    output_format = None
    transformation = "CAPITALISE"
    for o, a in options:
        if o in ("-i", "--input"):
            if a == "-":
                input_stream = sys.stdin
            else:
                filename = a
        if o in ("-o", "--output"):
            if a == "-":
                output_stream = sys.stdout
            else:
                output_file = a
        if o in ("-f", '--format'):
            input_format = a
        if o in ("-F", '--output_format'):
            output_format = a
        if o in ("-u", "--upper"):
            transformation = 'CAPITALISE'
        if o in ("-l", "--lower"):
            transformation = "DECAPITALISE"
    transform(filename, input_stream, input_format,
              output_format, transformation, output_file, output_stream)


if __name__ == "__main__":
    main()

File: transformer.go
package main

import (
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"strings"

	"gopkg.in/yaml.v2"
)

func main() {
	var filename string
	var stream io.Reader
	var inputFormat string
	var outputFormat string
	var transformation string
	var outputFile string
	var outputStream io.Writer

	flag.StringVar(&filename, "i", "", "File to read from, or '-' for STDIN")
	flag.StringVar(&inputFormat, "f", "", "Input format - JSON or YAML, if blank will be derived from filename")
	flag.StringVar(&outputFormat, "F", "", "Output format - JSON or YAML, if blank will be derived from filename")
	flag.StringVar(&transformation, "t", "CAPITALISE", "Transformation to apply")
	flag.StringVar(&outputFile, "o", "", "File to write to, or '-' for STDOUT")

	flag.Parse()

	if filename == "-" {
		stream = os.Stdin
		filename = ""
	}

	if outputFile == "-" {
		outputStream = os.Stdout
		outputFile = ""
	}

	if err := Transform(filename, stream, inputFormat, outputFormat, transformation, outputFile, outputStream); err != nil {
		fmt.Fprintln(os.Stderr, err.Error())
		os.Exit(1)
	}
}

func Transform(filename string, stream io.Reader, inputFormat, outputFormat, transformation, outputFile string, outputStream io.Writer) (err error) {
	stuff := make([]string, 0)
	var raw_stuff []byte

	if filename != "" {
		fp, err := os.Open(filename)
		if err != nil {
			return err
		}
		if inputFormat == "" {
			if strings.HasSuffix(filename, ".json") {
				inputFormat = "JSON"
			} else if strings.HasSuffix(filename, ".yaml") || strings.HasSuffix(filename, ".yml") {
				inputFormat = "YAML"
			}
			if inputFormat == "" {
				fp.Close()
				return errors.New("No input format was specified")
			}
		}
		if inputFormat == "YAML" {
			raw_stuff, err = ioutil.ReadAll(fp)
			if err == nil {
				err = yaml.Unmarshal(raw_stuff, &stuff)
			}
		} else {
			decoder := json.NewDecoder(fp)
			if err := decoder.Decode(&stuff); err != nil {
				return err
			}
		}
		fp.Close()
		if err != nil {
			return err
		}
	} else {
		raw_stuff, err = ioutil.ReadAll(stream)
		if inputFormat == "" {
			return errors.New("No input format was specified")
		}
		if inputFormat == "YAML" {
			if err := yaml.Unmarshal(raw_stuff, &stuff); err != nil {
				return err
			}
		} else if inputFormat == "JSON" {
			if err := json.Unmarshal(raw_stuff, &stuff); err != nil {
				return err
			}
		}
	}

	if outputFormat == "" {
		if strings.HasSuffix(outputFile, ".json") {
			outputFormat = "JSON"
		} else if strings.HasSuffix(outputFile, ".yaml") || strings.HasSuffix(outputFile, ".yml") {
			outputFormat = "YAML"
		} else {
			outputFormat = inputFormat
		}
	}
	more_stuff := make([]string, 0)

	if transformation == "CAPITALISE" {
		for _, element := range stuff {
			more_stuff = append(more_stuff, strings.ToUpper(element))
		}
	} else if transformation == "DECAPITALISE" {
		for _, element := range stuff {
			more_stuff = append(more_stuff, strings.ToLower(element))
		}
	}

	if outputFormat == "JSON" {
		if outputFile != "" {
			data, err := json.Marshal(&more_stuff)
			if err != nil {
				return err
			}
			if err := ioutil.WriteFile(outputFile, data, 0666); err != nil {
				return err
			}
		} else if outputStream != nil {
			encoder := json.NewEncoder(outputStream)
			encoder.Encode(&more_stuff)
		} else {
			return errors.New("No output stream is specified")
		}
	}

	if outputFormat == "YAML" {
		data, err := yaml.Marshal(&more_stuff)
		if err != nil {
			return err
		}
		if outputStream != nil {
			if _, err := outputStream.Write(data); err != nil {
				return err
			}
		} else {
			if err := ioutil.WriteFile(outputFile, data, 0666); err != nil {
				return err
			}
		}
	}

	return nil
}
